language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
             unfoldFixOf (Normalized vPat) (Normalized vArg) k
         _                  -> throwError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
 
+
+-- Typecheck a builtin application.  This performs a lot of
+-- substitutions and normalisations, but it should preserve the global
+-- uniqueness condition because it only uses uniquness-preserving
+-- operations from Normalize.Internal.
+
+-- [infer| bn : [a_1 :: K_1, ..., a_n :: K_m](B_1, ..., B_n)C]
+-- [check| G !- A_i :: K_i]
+-- [check| G !- M_j : T_j]     [A_i / a_i]* B_j ~>? S_j   T_j ~ S_j
```

Comment:
I think you directly check `M_j` against `S_j` without introducing any intermediate `T_j`?